### PR TITLE
fix: add dedupe capability on all IO operations

### PR DESCRIPTION
**Pull request by ssddOnTop**
Issue Reference(s):
Fixes https://github.com/tailcallhq/tailcall/issues/2893

### CODE REVIEW

File: src/core/jit/model.rs [hunk 1]
```diff
@@ -408,6 +410,14 @@ impl<Input> OperationPlan<Input> {
...
             .map(|f| f.into_nested(&fields))
             .collect::<Vec<_>>();
+        let dedupe = fields.iter().filter(|v| v.ir.is_some()).all(|v| {
+            v.ir.as_ref()
+                .map(|v| match v {
+                    IR::IO(io) => io.dedupe(),
+                    _ => false,
+                })
+                .unwrap_or_default()
+        });
```
Comment (tusharmath) [inline src/core/jit/model.rs:420]: Move this logic to a function and add a unit test.

File: src/core/jit/model.rs [hunk 2]
```diff
@@ -410,7 +410,7 @@ impl<Input> OperationPlan<Input> {
             .filter(|f| f.extensions.is_none())
             .map(|f| f.into_nested(&fields))
             .collect::<Vec<_>>();
-        let dedupe = fields.iter().filter(|v| v.ir.is_none()).all(|v| {
+        let dedupe = fields.iter().filter(|v| v.ir.is_some()).all(|v| {
```
Comment (tusharmath) [inline src/core/jit/model.rs:413]: We don't need `filter` Can pattern match directly inside `all` and return a true or false.

File: tailcall-fixtures/fixtures/configs/jsonplaceholder.graphql [hunk 3]
```diff
@@ -5,7 +5,7 @@ schema
 }
 
 type Query {
-  posts: [Post] @http(path: "/posts?_limit=11")
+  posts: [Post] @http(path: "/posts?_limit=11", dedupe: true)
```
Comment (tusharmath) [inline tailcall-fixtures/fixtures/configs/jsonplaceholder.graphql:8]: create a new test file `dedupe.graphql` with different `dedupe` settings.

File: src/core/jit/model.rs [hunk 4]
```diff
@@ -703,4 +714,29 @@ mod test {
...
+        let config = include_config!("../../../tailcall-fixtures/fixtures/configs/jsonplaceholder.graphql").unwrap();
+        let module = ConfigModule::from(config);
+        let bp = Blueprint::try_from(&module).unwrap();
+
+        let request = Request::new(r#"{ users { id } }"#);
+        let jit_request = jit::Request::from(request);
+        let plan = jit_request.create_plan(&bp).unwrap();
+
+        assert!(!plan.dedupe);
+    }
```
Comment (tusharmath) [inline src/core/jit/model.rs:743]: add another test where there are nested calls and some are deduped and some are not.

Comment (ssddOnTop) [reply src/core/jit/model.rs:743]: it's checked on flattened form of field

**Review by tusharmath — APPROVED** [review]

**Pull request merged**